zipkin/src/main/java/zipkin2/internal/ReadBuffer.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zipkin2.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+import static zipkin2.internal.HexCodec.HEX_DIGITS;
+import static zipkin2.internal.JsonCodec.UTF_8;
+
+/** Read operations do bounds checks, as typically more errors occur reading than writing. */
+public abstract class ReadBuffer extends InputStream {
+  /** Do not use the buffer passed here after, as it may be manipulated directly. */
+  public static ReadBuffer wrapUnsafe(ByteBuffer buffer) {
+    if (buffer.hasArray()) {
+      int pos = buffer.position();
+      return wrap(buffer.array(), buffer.arrayOffset() + pos);
+    }
+    return new ReadBuffer.Buff(buffer);
+  }
+
+  public static ReadBuffer wrap(byte[] bytes, int pos) {
+    return new ReadBuffer.Array(bytes, pos);
+  }
+
+  static final class Buff extends ReadBuffer {
+    final ByteBuffer buf; // visible for testing
+
+    Buff(ByteBuffer buf) {
+      this.buf = buf;
+    }
+
+    @Override final byte readByteUnsafe() {
+      return buf.get();
+    }
+
+    @Override final byte[] readBytes(int length) {
+      require(length);
+      byte[] copy = new byte[length];
+      buf.get(copy);
+      return copy;
+    }
+
+    @Override boolean tryReadAscii(char[] destination, int length) {
```

Comment:
dunno kindof doubt it because there is decoding going on in the CTOR as opposed to the char array one which does a plain copy. feel free to test it though

Ex from JDK 11
```
    public String(byte bytes[], int offset, int length) {
        checkBoundsOffCount(offset, length, bytes.length);
        StringCoding.Result ret = StringCoding.decode(bytes, offset, length);
        this.value = ret.value;
        this.coder = ret.coder;
    }
```
```